drv/ignition-server/src/main.rs

```diff
@@ -78,20 +81,95 @@ struct ServerImpl {
 }
 
 impl ServerImpl {
+    /// Poll the presence summary and track Targets arriving and departing.
     fn poll_presence(&mut self) -> Result<(), IgnitionError> {
         let current_presence_summary = self.controller.presence_summary()?;
 
         if current_presence_summary != self.last_presence_summary {
-            ringbuf_entry!(Trace::PresenceUpdate(current_presence_summary));
-            self.last_presence_summary = current_presence_summary;
+            let arriving_targets =
+                current_presence_summary & !self.last_presence_summary;
+            let departing_targets =
+                !current_presence_summary & self.last_presence_summary;
+
+            let arrived_targets = Self::map_ports(arriving_targets, |port| {
+                self.target_arrive(port)
+            });
+            let departed_targets = Self::map_ports(departing_targets, |port| {
+                self.target_depart(port)
+            });
+
+            // Update the presence summary based on targets which were
+            // succesfully processed. If a target wasn't processed it'll get
+            // retried on the next cycle.
+            self.last_presence_summary = arrived_targets
+                | (self.last_presence_summary & !departed_targets);
+
+            ringbuf_entry!(Trace::PresenceUpdate(self.last_presence_summary));
         }
 
         Ok(())
     }
 
+    /// Apply the given function to each port for which a bit in the `ports`
+    /// vector is set. Returns a bit vector with bits set for ports for which
+    /// the operation was succesfull. Under normal circumstances this output
+    /// vector is expected to match the input vector.
+    fn map_ports<F>(mut ports: u64, mut f: F) -> u64
+    where
+        F: FnMut(u8) -> Result<(), IgnitionError>,
+    {
+        let mut success = 0u64;
+
+        for port in 0..PORT_MAX as u8 {
+            if ports & 0x1 != 0 {
+                match f(port) {
+                    Ok(()) => success |= 1 << PORT_MAX,
```

Comment:
This sets the same bit each time, I think you mean:
```suggestion
                    Ok(()) => success |= 1 << port,
```